### PR TITLE
feat(Core/Mail): Server mail

### PR DESCRIPTION
**Pull request by Kitzunu**
## Changes Proposed:
-  Introduce mail templates to be set and sent at player login. Kind of how blizzard handles anniversary gifts.

## Tests Performed:

- works ingame

## How to Test the Changes:

1. Populate mail_server_template table with something.
2. Login, see that if the character meets the requirements, you will get the mail
3. Relog and see that you will not get the mail again since it was registered in mail_server_character

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/scripts/World/server_mail.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+public:
+    ServerMailReward() : PlayerScript("ServerMailReward") { }
+
+    // CHARACTER_LOGIN = 8
+    void OnLogin(Player* player) override
+    {
+        for (auto const& servMail : sObjectMgr->GetAllServerMailStore())
+        {
+            if (CharacterDatabase.Query("SELECT mailId from mail_server_character where guid = {} and mailId = {}", player->GetGUID().GetCounter(), servMail.second.id))
+                continue;
```
[the remark this answers is not in the document]
Comment (Kitzunu) [reply src/server/scripts/World/server_mail.cpp:33]: But that wont work either...

File: src/server/game/Globals/ObjectMgr.cpp [hunk 2]
```diff
@@ -9761,3 +9761,116 @@ uint32 ObjectMgr::GetQuestMoneyReward(uint8 level, uint32 questMoneyDifficulty)
...
+        if (player->GetTotalPlayedTime() < reqPlayTime)
+            return;
+
+        CharacterDatabaseTransaction trans = CharacterDatabase.BeginTransaction();
+
+        MailSender sender(MAIL_NORMAL, player->GetGUID().GetCounter(), MAIL_STATIONERY_GM);
+        MailDraft draft(subject, body);
+
+        draft.AddMoney(player->GetTeamId() == ALLIANCE ? rewardMoneyH : rewardMoneyA);
+        if (Item* mailItem = Item::CreateItem(player->GetTeamId() == ALLIANCE ? rewardItemH : rewardItemA, player->GetTeamId() == ALLIANCE ? rewardItemCountH : rewardItemCountA))
```
Comment (Kitzunu) [inline src/server/game/Globals/ObjectMgr.cpp:9781]:
```suggestion
        if (Item* mailItem = Item::CreateItem(player->GetTeamId() == TEAM_ALLIANCE ? rewardItemH : rewardItemA, player->GetTeamId() == TEAM_ALLIANCE ? rewardItemCountH : rewardItemCountA))
```

File: src/server/game/Globals/ObjectMgr.cpp [hunk 3]
```diff
@@ -9761,3 +9761,116 @@ uint32 ObjectMgr::GetQuestMoneyReward(uint8 level, uint32 questMoneyDifficulty)
...
+
+        if (player->GetTotalPlayedTime() < reqPlayTime)
+            return;
+
+        CharacterDatabaseTransaction trans = CharacterDatabase.BeginTransaction();
+
+        MailSender sender(MAIL_NORMAL, player->GetGUID().GetCounter(), MAIL_STATIONERY_GM);
+        MailDraft draft(subject, body);
+
+        draft.AddMoney(player->GetTeamId() == ALLIANCE ? rewardMoneyH : rewardMoneyA);
```
Comment (Kitzunu) [inline src/server/game/Globals/ObjectMgr.cpp:9780]:
```suggestion
        draft.AddMoney(player->GetTeamId() == TEAM_ALLIANCE ? rewardMoneyH : rewardMoneyA);
```

File: src/server/game/Mails/Mail.h [hunk 4]
```diff
@@ -211,4 +211,21 @@ struct Mail
...
+    uint8 reqLevel{ 0 };
+    uint32 reqPlayTime{ 0 };
+    uint32 moneyA{ 0 };
+    uint32 moneyH{ 0 };
+    uint32 itemA{ 0 };
+    uint32 itemCountA{ 0 };
+    uint32 itemH{ 0 };
+    uint32 itemCountH{ 0 };
+    std::string subject{ "" };
+    std::string body{ "" };
```
Comment (Winfidonarleyan) [inline src/server/game/Mails/Mail.h:227]: this empty default

File: src/server/scripts/World/server_mail.cpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            WorldSession* mySess = player->GetSession();
+            mySess->GetQueryProcessor().AddCallback(CharacterDatabase.AsyncQuery(stmt)
+                .WithPreparedCallback([mySess](PreparedQueryResult result)
+                    {
+                        for (auto const& servMail : sObjectMgr->GetAllServerMailStore())
+                        {
+                            if (result)
+                                continue;
+
+                            sObjectMgr->SendServerMail(mySess->GetPlayer(), servMail.second.id, servMail.second.reqLevel, servMail.second.reqPlayTime, servMail.second.moneyA, servMail.second.moneyH, servMail.second.itemA, servMail.second.itemCountA, servMail.second.itemH, servMail.second.itemCountH, servMail.second.subject, servMail.second.body, servMail.second.active);
```
Comment (Winfidonarleyan) [inline src/server/scripts/World/server_mail.cpp:47]: use send in 2-nd argument `struct ServerMail`

Comment (Kitzunu) [reply src/server/scripts/World/server_mail.cpp:47]: I dont understand

**Comment by Lynethris** [comment]
Tested and work. (Done with:  INSERT INTO 'mail_server_template' VALUES (1, 1, 0, 100, 1000, 17705, 1, 0, 0, "Good start", "Start", 1);
- [ ] Player recive a mail from client assistance.
- [ ] All values from the DB work
- [ ] Player get only one time the mail
- [ ] Command .reload mail_server_template work

Information:
-  Command .reload mail_server_template will make that all players will receive all mails (even if players have already get one of the mails)
- If a player recive the mail during this game session, the mail doesn't show up. He need to relog (example: set reqLevel to 40, create a char level 1 and use .char level 40. You will need to relog for see the mail.)

**Comment by Kitzunu** [comment]
> Command .reload mail_server_template will make that all players will receive all mails (even if players have already get one of the mails)

I will look into this

> If a player recive the mail during this game session, the mail doesn't show up. He need to relog (example: set reqLevel to 40, create a char level 1 and use .char level 40. You will need to relog for see the mail.)

This is intended

**Comment by Kitzunu** [comment]
> Command .reload mail_server_template will make that all players will receive all mails (even if players have already get one of the mails)

@Lynethris I cannot reproduce this. For me it all works as it should

File: src/server/game/Mails/Mail.h (hunk 4 above)
Comment (Kitzunu) [reply src/server/game/Mails/Mail.h:227]: Only need `std::string body{ }` ?

**Comment by Lynethris** [comment]
My step: 
- Add one thing in the DB. 
- Start server and log on char
- Get mail
- While server running add a second thing in the DB
- IG use command and relog

Normally, you get the 2nd and the first mail (even though we already had it)

**Comment by Kitzunu** [comment]
Okay, I got to reproduce it now. Hmm... Thanks

File: src/server/scripts/World/server_mail.cpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            stmt->SetData(1, servMail.second.id);
+
+            WorldSession* mySess = player->GetSession();
+            mySess->GetQueryProcessor().AddCallback(CharacterDatabase.AsyncQuery(stmt)
+                .WithPreparedCallback([mySess](PreparedQueryResult result)
+                    {
+                        for (auto const& servMail : sObjectMgr->GetAllServerMailStore())
+                        {
+                            if (result)
+                                continue;
```
Comment (Kitzunu) [inline src/server/scripts/World/server_mail.cpp:45]: I really do not understand how this always works. But is skipped if you do .reload mail_server_template

even though all data is the same in DB?

Comment (Kitzunu) [reply src/server/scripts/World/server_mail.cpp:45]: It only happens when you reload after adding a new entry. not everytime you reload.

Could it have something to do with the server mail store? unsure

File: src/server/scripts/World/server_mail.cpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        for (auto const& servMail : sObjectMgr->GetAllServerMailStore())
+        {
+            CharacterDatabasePreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_SEL_MAIL_SERVER_CHARACTER);
+            stmt->SetData(0, player->GetGUID().GetCounter());
+            stmt->SetData(1, servMail.second.id);
+
+            WorldSession* mySess = player->GetSession();
+            mySess->GetQueryProcessor().AddCallback(CharacterDatabase.AsyncQuery(stmt)
+                .WithPreparedCallback([mySess](PreparedQueryResult result)
```
Comment (Kitzunu) [inline src/server/scripts/World/server_mail.cpp:40]:
```suggestion
                .WithPreparedCallback([mySess, servMail](PreparedQueryResult result)
```

File: src/server/scripts/World/server_mail.cpp [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                    {
+                        for (auto const& servMail : sObjectMgr->GetAllServerMailStore())
```
Comment (Kitzunu) [inline src/server/scripts/World/server_mail.cpp:42]:
```suggestion
```

**Comment by Kitzunu** [comment]
@Lynethris try again please

File: src/server/game/Mails/Mail.h (hunk 4 above)
Comment (Winfidonarleyan) [reply src/server/game/Mails/Mail.h:227]:
```cpp
std::string body
```

Comment (Winfidonarleyan) [reply src/server/game/Mails/Mail.h:227]: image

**Comment by Lynethris** [comment]
@Kitzunu

Screenshot 2022-04-19 084541

File: src/server/scripts/World/server_mail.cpp [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+            WorldSession* mySess = player->GetSession();
+            mySess->GetQueryProcessor().AddCallback(CharacterDatabase.AsyncQuery(stmt)
+                .WithPreparedCallback([mySess, servMail](PreparedQueryResult result)
+                    {
+                        {
+                            if (result)
+                                continue;
+
+                            sObjectMgr->SendServerMail(mySess->GetPlayer(), servMail.second.id, servMail.second.reqLevel, servMail.second.reqPlayTime, servMail.second.moneyA, servMail.second.moneyH, servMail.second.itemA, servMail.second.itemCountA, servMail.second.itemH, servMail.second.itemCountH, servMail.second.subject, servMail.second.body, servMail.second.active);
```
Comment (Kitzunu) [inline src/server/scripts/World/server_mail.cpp:46]:
```suggestion
                                 sObjectMgr->SendServerMail(mySess->GetPlayer(), servMail.second.id, servMail.second.reqLevel, servMail.second.reqPlayTime, servMail.second.moneyA, servMail.second.moneyH, servMail.second.itemA, servMail.second.itemCountA, servMail.second.itemH, servMail.second.itemCountH, servMail.second.subject, servMail.second.body, servMail.second.active);
```

File: src/server/scripts/World/server_mail.cpp [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+            stmt->SetData(1, servMail.second.id);
+
+            WorldSession* mySess = player->GetSession();
+            mySess->GetQueryProcessor().AddCallback(CharacterDatabase.AsyncQuery(stmt)
+                .WithPreparedCallback([mySess, servMail](PreparedQueryResult result)
+                    {
+                        {
+                            if (result)
+                                continue;
+
```
Comment (Kitzunu) [inline src/server/scripts/World/server_mail.cpp:45]:
```suggestion
```

File: src/server/scripts/World/server_mail.cpp [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+            stmt->SetData(0, player->GetGUID().GetCounter());
+            stmt->SetData(1, servMail.second.id);
+
+            WorldSession* mySess = player->GetSession();
+            mySess->GetQueryProcessor().AddCallback(CharacterDatabase.AsyncQuery(stmt)
+                .WithPreparedCallback([mySess, servMail](PreparedQueryResult result)
+                    {
+                        {
+                            if (result)
+                                continue;
```
Comment (Kitzunu) [inline src/server/scripts/World/server_mail.cpp:44]:
```suggestion
```

File: src/server/scripts/World/server_mail.cpp [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+            CharacterDatabasePreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_SEL_MAIL_SERVER_CHARACTER);
+            stmt->SetData(0, player->GetGUID().GetCounter());
+            stmt->SetData(1, servMail.second.id);
+
+            WorldSession* mySess = player->GetSession();
+            mySess->GetQueryProcessor().AddCallback(CharacterDatabase.AsyncQuery(stmt)
+                .WithPreparedCallback([mySess, servMail](PreparedQueryResult result)
+                    {
+                        {
+                            if (result)
```
Comment (Kitzunu) [inline src/server/scripts/World/server_mail.cpp:43]:
```suggestion
                            if (!result)
```

**Comment by Kitzunu** [comment]
@Lynethris try now, doing it via phone lmao

**Comment by Lynethris** [comment]
@Kitzunu It's work fine.

File: src/server/scripts/World/server_mail.cpp [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+            stmt->SetData(0, player->GetGUID().GetCounter());
+            stmt->SetData(1, servMail.second.id);
+
+            WorldSession* mySess = player->GetSession();
+            mySess->GetQueryProcessor().AddCallback(CharacterDatabase.AsyncQuery(stmt)
+                .WithPreparedCallback([mySess, servMail](PreparedQueryResult result)
+                    {
+                        {
+                            if (!result)
+                                 sObjectMgr->SendServerMail(mySess->GetPlayer(), servMail.second.id, servMail.second.reqLevel, servMail.second.reqPlayTime, servMail.second.moneyA, servMail.second.moneyH, servMail.second.itemA, servMail.second.itemCountA, servMail.second.itemH, servMail.second.itemCountH, servMail.second.subject, servMail.second.body, servMail.second.active);
```
Comment (Kitzunu) [inline src/server/scripts/World/server_mail.cpp:44]:
```suggestion
                                sObjectMgr->SendServerMail(mySess->GetPlayer(), servMail.second.id, servMail.second.reqLevel, servMail.second.reqPlayTime, servMail.second.moneyA, servMail.second.moneyH, servMail.second.itemA, servMail.second.itemCountA, servMail.second.itemH, servMail.second.itemCountH, servMail.second.subject, servMail.second.body, servMail.second.active);
```

File: src/server/game/Mails/Mail.h (hunk 4 above)
Comment (Kitzunu) [inline src/server/game/Mails/Mail.h:227]:
```suggestion
    std::string subject;
    std::string body;
```

**Review by Nyeriah — APPROVED** [review]

**Comment by mpfans** [comment]
wiki need update..

**Pull request merged**